BridgeSDK/BridgeAPI/SBBUploadManager.m

```diff
@@ -255,6 +255,125 @@ - (SBBUploadSession *)uploadSessionForFile:(NSString *)fileURLString
 }
 
+// When uploading a file to Bridge, first we use background download to get an SBBUploadSession object from Bridge.
+// Then we use the pre-signed URL in that SBBUploadSession to upload our file to S3.
+// When that finishes, we tell Bridge that the requested upload is completed.
+// To avoid having multiple layers of nested completion handlers in one method, they're broken out by step:
+
+// --- Here's the completion handler for the step where we tell Bridge the upload completed:
+- (void)toldBridgeFileUploadedWithTask:(NSURLSessionUploadTask *)uploadTask forBridgeSession:(SBBUploadSession *)uploadSession error:(NSError *)error
+{
+#if DEBUG
+    if (error) {
+        NSLog(@"Error calling upload complete for upload ID %@:\n%@", uploadSession.id, error);
+    } else {
+        NSString* uploadStatusUrlString = nil;
+        if ([self.networkManager isKindOfClass:[SBBNetworkManager class]]) {
+            NSString* relativeUrl = [NSString stringWithFormat:kSBBUploadStatusAPIFormat, uploadSession.id];
+            NSURL* url = [(SBBNetworkManager*) self.networkManager URLForRelativeorAbsoluteURLString:relativeUrl];
+            uploadStatusUrlString = [url absoluteString];
+            if ([_uploadDelegate respondsToSelector:@selector(uploadManager:uploadOfFile:completedWithVerificationURL:)]) {
+                [_uploadDelegate uploadManager:self uploadOfFile:uploadTask.taskDescription completedWithVerificationURL:url];
+            }
+        }
+        NSLog(@"Successfully called upload complete for upload ID %@, check status with curl -H \"Bridge-Session:%@\" %@", uploadSession.id, [self.authManager.authDelegate sessionTokenForAuthManager:self.authManager], uploadStatusUrlString);
+    }
+#endif
+    [self completeUploadOfFile:uploadTask.taskDescription withError:error];
+}
+
+// --- Here's the completion handler for the step where we upload the file to S3:
+- (void)uploadedFileToS3WithTask:(NSURLSessionUploadTask *)uploadTask response:(NSHTTPURLResponse *)httpResponse error:(NSError *)error
+{
+    NSInteger httpStatusCode = httpResponse.statusCode;
+    
+    // client-side networking issue
+    if (error) {
+        [self completeUploadOfFile:uploadTask.taskDescription withError:error];
+        return;
+    }
+    
+    // server didn't like the request, or otherwise hiccupped
+    if (httpStatusCode >= 300) {
+        // iOS handles redirects automatically so only e.g. 307 resource not changed etc. from the 300 range should end up here
+        // (along with all 4xx and 5xx of course)
+        NSString *description = [NSString stringWithFormat:@"Background file upload to S3 failed with HTTP status %ld", (long)httpStatusCode];
+        NSError *s3Error = [NSError errorWithDomain:SBB_ERROR_DOMAIN code:SBBErrorCodeS3UploadErrorResponse userInfo:@{NSLocalizedDescriptionKey: description}];
+        [self completeUploadOfFile:uploadTask.taskDescription withError:s3Error];
+        return;
+    }
+    
+    // tell the API we done did it
```

Comment:
moved this code here from where it used to be, mostly unchanged (including this comment)--probably had been reading or watching a lot of something with Southern or country dialect at the time? dunno.